### PR TITLE
EN Nearby api implementation on top of PPCP

### LINKED ISSUE #33: Make the API control EN API
Make the actual EN API control the internal parts of our PPCP implementation. The API calls should not just be a Mock but a full working implementation. Meaning it can trigger the BLE server/client and ask the exposure algorithm what the propability for corona exposure is.

CAUTION: This requires https://github.com/CoraLibre/CoraLibre-android-sdk/issues/34 to be implemented.

### PR DESCRIPTION
**Pull request by Stypox**
This PR makes the Nearby api actually do something, implementing the stub methods. I created a branch here so that you can easily push changes, too, in case you want to. I took inspiration from [microg's implementation](https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt).

Closes #33 

This is what needs to be done:

- [x] Rename ENNumber to ENInterval (the ENNumber name was not meaningful)
<br>

- [ ] **Full diagnosis key handling and matching**
- [x] Add `DiagnosisKey` enitity to database
- [ ] Add database tests for diagnosis keys
- [x] Add protobuf to decode key files (`.proto` file [taken from microg](https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-nearby-core-proto/src/main/proto/TemporaryExposureKeyFile.proto))
- [ ] Implement `findAllMeasuredExposures()` to match stored `CapturedData`s with stored `DiagnosisKey`s.
- [ ] Add database tests for `findAllMeasuredExposures()`
<br>

- [x] Link `start()`/`stop()` to `PPCP.start`/`.stop`
- [x] Implement checking if EN service `isEnabled()` (I added a private `boolean` for this, is it ok or shall I use `PPCP.isStarted()` for this instead?
- [x] Implement `getTemporaryExposureKeyHistory()`, returning all of the keys generated by the phone in the last 14 days
<br>

- [x] **Implement `getTemporaryExposureKeyHistory()`**
- [ ] Make sure TEK key data is correct
<br>

- [ ] **Implement `provideDiagnosisKeys()`**
- [x] Read Zip files and validate them ([mostly taken form microg](https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt#L139))
- [x] Store `DiagnosisKey`s in the database
- [x] Send intent to notify whether there was a measured exposure or not (like in microg)
- [ ] Verify whether we need to precalculate database matches or not
<br>

- [ ] **Implement `getExposureSummary()`**
- [x] Loop through matched keys and collect information
- [x] Calculate days from first capture timestamp
- [x] Calculate attenuations and attenuation durations (i.e. the time the user has been in contact with a particular attenuation level)
<br>

- [ ] **Implement `getExposureInformation()`**

### CODE REVIEW

File: sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java [hunk 1]
```diff
@@ -1,67 +1,182 @@
...
-        // TODO: Implement!
-        return Tasks.forResult(Collections.emptyList());
+        return Tasks.call(() -> {
+            Iterator<IntervalOfCapturedData> collectedPayload
+                = database.getAllCollectedPayload().iterator();
+            List<TemporaryExposureKey> result = new ArrayList<>();
+
+            while (collectedPayload.hasNext()) {
+                IntervalOfCapturedData intervalOfCapturedData = collectedPayload.next();
+                CapturedData firstCapturedData = intervalOfCapturedData.getCapturedData().get(0);
```
Comment (BjoernPetersen) [inline sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java:88]: Just FYI: there's need to manually iterate over an `Iterable`, you can use a for-each-loop:

```suggestion
            Iterable<IntervalOfCapturedData> collectedPayload = database.getAllCollectedPayload();
            List<TemporaryExposureKey> result = new ArrayList<>();

            for (IntervalOfCapturedData intervalOfCapturedData : collectedPayload) {
                CapturedData firstCapturedData = intervalOfCapturedData.getCapturedData().get(0);
```

File: sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java [hunk 2]
```diff
@@ -1,67 +1,182 @@
...
     }
 
     @Override
     public Task<Void> provideDiagnosisKeys(final List<File> keyFiles,
                                            @Nullable final ExposureConfiguration exposureConfiguration,
                                            final String token) {
-        // TODO: Implement!
-        return Tasks.forResult(null);
+        return Tasks.call(() -> {
+            // TODO save exposure configuration to database, to restore it after e.g. phone restart, like microg?
```
Comment (BjoernPetersen) [inline sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java:109]: Good idea, but I think it's fine to leave this as a TODO for now.

**Review by BjoernPetersen — COMMENTED** [review]
Looking good so far. As you know I'm in the process of fixing/implementing some of the `nearby.*` classes (#56), that's probably going to block you a little. BTW: When I'm done with that, some of those long-to-int casts are going to be unnecessary.

File: sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java (hunk 1 above)
Comment (Stypox) [reply sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java]: Done

**Comment by Stypox** [comment]
I pushed three more commits. I implemented diagnosis keys protobuf parsing using [microg's `.proto`](https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-nearby-core-proto/src/main/proto/TemporaryExposureKeyFile.proto), along with the database methods to store them, and I added `findAllMeasuredExposures()`.
But I am facing some problems matching the `CapturedData`s stored in the database with the provided `DiagnosisKey`s: the microg implementation is obscure for me, and I have some doubts about the `byte[]` stored in the `CapturedData` object. @theScrabi does that array represent just the `TemporaryExposureKey` received from the other phone, or does it represent the received data as a whole?
Also, I have a design question: I noticed that [microg caches matches](https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt#L192) inside `provideDiagnosisKeys`, so that they don't have to be recalculated later. Is it ok if for now I don't implement that (as it would require adding another table to the database), and then we'll get to it later?

**Comment by haitrec** [comment]
I implemented a big part of the actual matching procedure / computation of ExposureSummary and ExposureInformation. I did not write tests for these parts yet. Also there are some parts of the computation, where I was not 100% sure, I marked thse parts with TODOs in the code.
I did some restructuring along the way, in particular I moved all datatypes that are used in different parts of the sdk to an own 'datatypes' package.
When reviewing, better don't look at every change from every commit, but on the result, since not all commits are super clean ;)

File: sdk/src/main/java/org/coralibre/android/sdk/internal/TracingService.java [hunk 3]
```diff
@@ -57,7 +57,8 @@
     // The source of these Values is this a description on Googles API description:
     // https://developers.google.com/android/exposure-notifications/ble-attenuation-overview
     public static final long SCAN_DURATION = 4 * 1000; // unit in milliseconds
-    public static final long SCAN_INTERVAL = 5 * 60 * 1000; // unit in milliseconds
+    public static final long SCAN_INTERVAL_SECONDS = 5 * 60; // unit in milliseconds
```
Comment (Stypox) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/TracingService.java:60]: unit in seconds, not milliseconds

File: sdk/src/main/java/org/coralibre/android/sdk/internal/database/persistent/DaoExposureInformation.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.coralibre.android.sdk.internal.database.persistent.entity.EntityExposureInformation;
+
+import java.util.List;
+
+@Dao
+public interface DaoExposureInformation {
+
+    @Insert(onConflict = OnConflictStrategy.ABORT)
+    public void insertExposureInformations(List<EntityExposureInformation> entityExposureInformations);
```
Comment (Stypox) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/database/persistent/DaoExposureInformation.java:16]: I don't think "informations" is a valid word in English, but nevermind, ignore this ;-)

File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package org.coralibre.android.sdk.internal.datatypes;
+
+import java.util.List;
+
+public interface IntervalOfCapturedData {
```
Comment (Stypox) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java:5]: Maybe this interface can be removed? It has (and will always have) only one implementation

Comment (Stypox) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java:5]: Maybe this interface can be removed? It has (and will always have) only one implementation

File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+    ENInterval getInterval();
+
+    List<CapturedData> getCapturedData();
+
+    void add(final CapturedData capturedData);
+
+}
```
Comment (Stypox) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java:13]: Maybe this interface can be removed? It has (and will always have) only one implementation

File: sdk/src/main/java/org/coralibre/android/sdk/internal/matching/ExposureUtils.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        // Now assemble Exposure objects from consecutive matches that have only a small difference
+        // in the capture time:
+
+        // ... for that, we first split the match list to get one list per exposure:
+        LinkedList<LinkedList<Match>> matchesPerExposure = new LinkedList<>();
+        final long lastMatchTimestampMillis = -1;
+        for (Match match : matchesSorted) {
+            if (lastMatchTimestampMillis == -1
+                || (match.captureTimestampMillis - lastMatchTimestampMillis) * 1000
+                > EnFrameworkConstants.MAX_EXPOSURE_INTERPOLATION_DURATION_SECONDS) {
```
Comment (Stypox) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/matching/ExposureUtils.java:67]: This looks strange to me: a microseconds variable (it was millis, then by multiplying by 1000 it became micros) is compared to a seconds one.

**Review by Stypox — COMMENTED** [review]
The new structure looks far clearer than before to me, thank you for figuring it out! I didn't look thouroughly at the changes, as they are far too many and I don't have much time, but I agree this should be merged even if not ready and not perfect, so that people can start contributing on smaller parts of code. Thank you :-D

File: sdk/src/main/java/org/coralibre/android/sdk/internal/matching/ExposureUtils.java (hunk 7 above)
Comment (haitrec) [reply sdk/src/main/java/org/coralibre/android/sdk/internal/matching/ExposureUtils.java]: Fixed

File: sdk/src/main/java/org/coralibre/android/sdk/internal/TracingService.java (hunk 3 above)
Comment (haitrec) [reply sdk/src/main/java/org/coralibre/android/sdk/internal/TracingService.java]: I changed the names and removed the comments, ty

File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java (hunk 6 above)
Comment (haitrec) [reply sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/IntervalOfCapturedData.java]: I think you are right, I deleted it now. If this was a mistake, it shouldn't be too hard to revert.

**Review by Stypox — COMMENTED** [review]
I can't approve because I initially created the PR myself, but this is ready and can be merged, and as you said we will take care of other things later ;-)
Thank you :-D

**Comment by BjoernPetersen** [comment]
I'll review this tomorrow 👍

File: sdk/src/main/java/org/coralibre/android/sdk/internal/bluetooth/BleClient.java [hunk 8]
```diff
@@ -149,11 +149,11 @@ public void onScanFailed(int errorCode) {
     private void onDeviceFound(ScanResult scanResult) {
         try {
             long now = System.currentTimeMillis();
-            ENNumber currentInterval = CryptoModule.getCurrentInterval();
+            ENInterval currentInterval = ENIntervalUtil.getCurrentInterval();
             BluetoothDevice bluetoothDevice = scanResult.getDevice();
             final String deviceAddr = bluetoothDevice.getAddress();
 
-            byte rssi = (byte) scanResult.getRssi();
+            byte rssi = new Integer(scanResult.getRssi()).byteValue();
```
Comment (BjoernPetersen) [inline sdk/src/main/java/org/coralibre/android/sdk/internal/bluetooth/BleClient.java:156]: In general, you should never use the Integer constructor. Additionally, boxing a primitive int to truncate it to a byte is not necessary. Casting the int to byte gives you the same result more efficiently.

```suggestion
            byte rssi = (byte) scanResult.getRssi();
```

**Review by BjoernPetersen — CHANGES_REQUESTED** [review]

**Comment by BjoernPetersen** [comment]
I made a few rather minor suggestions. When they are fixed/applied, I'd be okay with merging this PR.

**Comment by BjoernPetersen** [comment]
I'll apply the suggested changes myself and merge this PR.

**Review by BjoernPetersen — APPROVED** [review]

**Pull request merged**